hyperspy/learn/rpca.py

```diff
@@ -0,0 +1,389 @@
+    m, n = X.shape
+    s = np.zeros(m)
+    x = np.zeros(n)
+    maxiter = 1e9
+    itr = 0
+
+    ddt = np.dot(scipy.linalg.inv(np.dot(X.T, X) + I), X.T)
+
+    while True:
+        itr += 1
+        xtmp = x
+        x = np.dot(ddt, (z - s))
+        stmp = s
+        s = np.maximum(z - np.dot(X, x) - lambda2, 0.0)
+        stopx = np.sqrt(np.dot(x - xtmp, (x - xtmp).conj()))
+        stops = np.sqrt(np.dot(s - stmp, (s - stmp).conj()))
+        stop = max(stopx, stops) / m
+        if stop < 1e-6 or itr > maxiter:
+            break
+
+    return x, s
+
+def _updatecol(X, A, B, I):
+    tmp, n = X.shape
+    L = X
+    A = A + I
+
+    for i in range(n):
+        b = B[:, i]
+        x = X[:, i]
+        a = A[:, i]
+        temp = (b - np.dot(X, a)) / A[i, i] + x
+        L[:, i] = temp / max(np.sqrt(np.dot(temp, temp.conj())), 1)
+
+    return L
+
+def orpca(X, rank, fast=False,
+          lambda1=None,
+          lambda2=None,
+          method=None,
+          learning_rate=None,
+          init=None,
+          training_samples=None):
+    """
+    This function performs Online Robust PCA
+    with missing or corrupted data.
+
+    Parameters
+    ----------
+    X : numpy array
+        is the [nfeatures x nsamples] matrix of observations.
+    rank : int
+        The model dimensionality.
+    lambda1 : None | float
+        Nuclear norm regularization parameter.
+        If None, set to 1 / sqrt(nsamples)
+    lambda2 : None | float
+        Sparse error regularization parameter.
+        If None, set to 1 / sqrt(nsamples)
+    method : None | 'CF' | 'BCD' | 'SGD'
+        'CF'  - Closed-form solver
+        'BCD' - Block-coordinate descent
+        'SGD' - Stochastic gradient descent
+        If None, set to 'CF'
+    learning_rate : None | float
+        Learning rate for the stochastic gradient
+        descent algorithm
+        If None, set to 1
+    init : None | 'qr' | 'rand'
+        'qr'   - QR-based initialization
+        'rand' - Random initialization
+        If None, set to 'qr'
+    training_samples : integer
+        Specifies the number of training samples to use in
+        the 'qr' initialization
+        If None, set to 10
+
+    Returns
+    -------
+    Xhat : numpy array
+        is the [nfeatures x nsamples] low-rank matrix
+    Ehat : numpy array
+        is the [nfeatures x nsamples] sparse error matrix
+    U, S, V : numpy arrays
+        are the results of an SVD on Xhat
+
+    Notes
+    -----
+    The ORPCA code is based on a transcription of MATLAB code obtained from
+    the following research paper:
+       Jiashi Feng, Huan Xu and Shuicheng Yuan, "Online Robust PCA via
+       Stochastic Optimization", Advances in Neural Information Processing
+       Systems 26, (2013), pp. 404-412.
+
+    It has been updated to include a new initialization method based
+    on a QR decomposition of the first n "training" samples of the data.
+    A stochastic gradient descent solver is also implemented.
+
```

Comment:
If this is a translation from MATLAB code, could you provide a link to the original code?